plugin-statusnotifier/statusnotifierproxy.cpp

```diff
@@ -0,0 +1,129 @@
+ * Copyright: 2021 LXQt team
+ *
+ * This program or library is free software; you can redistribute it
+ * and/or modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General
+ * Public License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA
+ *
+ * END_COMMON_COPYRIGHT_HEADER */
+
+#include "statusnotifierproxy.h"
+#include "statusnotifierwatcher.h"
+#include "statusnotifierwidget.h"
+
+#include <QApplication>
+#include <QFutureWatcher>
+#include <QtConcurrent>
+#include <QDBusConnectionInterface>
+#include <QDebug>
+
+Q_GLOBAL_STATIC(StatusNotifierProxy, statusNotifierProxy)
+
+StatusNotifierProxy::StatusNotifierProxy()
+    : mWatcher(nullptr),
+    mWidgetCount(0),
+    mCreatingWatcher(false)
+{
+    createWatcher();
+}
+
+void StatusNotifierProxy::createWatcher()
+{
+    mCreatingWatcher = true;
+    QFutureWatcher<StatusNotifierWatcher *> * future_watcher = new QFutureWatcher<StatusNotifierWatcher *>;
+    connect(future_watcher, &QFutureWatcher<StatusNotifierWatcher *>::finished, this, [this, future_watcher]
+        {
+            mWatcher = future_watcher->future().result();
+            watcherCreated();
+
+            connect(mWatcher, &StatusNotifierWatcher::StatusNotifierItemRegistered,
+                    this, &StatusNotifierProxy::onStatusNotifierItemRegistered);
+            connect(mWatcher, &StatusNotifierWatcher::StatusNotifierItemUnregistered,
+                    this, &StatusNotifierProxy::onStatusNotifierItemUnregistered);
+
+            qDebug() << mWatcher->RegisteredStatusNotifierItems();
+
+            future_watcher->deleteLater();
+        });
+
+    QFuture<StatusNotifierWatcher *> future = QtConcurrent::run([]
+        {
+            QString dbusName = QStringLiteral("org.kde.StatusNotifierHost-%1-%2").arg(QApplication::applicationPid()).arg(1);
+            if (QDBusConnectionInterface::ServiceNotRegistered == QDBusConnection::sessionBus().interface()->registerService(dbusName, QDBusConnectionInterface::DontQueueService))
+                qDebug() << "unable to register service for " << dbusName;
+
+            StatusNotifierWatcher * watcher = new StatusNotifierWatcher;
+            watcher->RegisterStatusNotifierHost(dbusName);
+            watcher->moveToThread(QApplication::instance()->thread());
+            return watcher;
+        });
+
+    future_watcher->setFuture(future);
+}
+
+void StatusNotifierProxy::watcherCreated()
+{
+    mCreatingWatcher = false;
```

Comment:
Isn't the `mCreatingWatcher` pointless, if we're relying on the nullity of `mWatcher` anyway?

Reply:
It was supposed to prevent it being created again in the time between creating the `QFutureWatcher` and it being resolved.
I didn't see it happen, but i think it could happen when starting `lxqt-panel` and there are multiple panels.